### DISCUSSION TITLE
permission to copy course content

### DISCUSSION THREAD

**Discussion by Alex-Jordan** [openwebwork/webwork2#2353] [General]
I'd like to make it possible to be logged into course A as "professor", and with a few clicks import the files and exercise sets from course B. This can be accomplished now, but with the following barriers:

1. Course B must be active (not archived and closed).
2. You have an account in course B, and it has high enough permissions to export sets and access the file manager.
4. You have to log in to course B, archive exercise sets, download the set def files plus any other file content
5. You have to log in to course A, upload it all, and import the set def files.

But I just want to be in course A, select course B from a list (even if course B is archived and closed), click "Take Action" and voila.

Let's make the huge assumption that I get through all the technical barriers to make that doable. It would be wrong to just allow people to copy whatever from any course on the WW server. You should have to have permission to do this. So my question for discussion is what would qualify for permission? In particular, is it enough for there to just be a professor-level account in course B with the same username as your user in course A? Or would some new structure be needed to manage "permission to copy course content"?

**Comment by dlglin** [comment]
This gets complicated due to our lack of global users.  There is no way to be certain that the same username across multiple courses belongs to the same person.  In a single-institution WW installation it's very likely that this will be the case, but I can't think of a reliable way to confirm.

Some random thoughts:

One way to make sure that the user has appropriate access in course B is to force them to authenticate in that course.  This could be the first step in the copy process, or it could be a separate action (e.g. "link another course") that is done beforehand.  My initial inclination is toward the former.  When I copy stuff from one course to another I typically do it all at once, so setting up a persistent relationship between the courses would be extra steps that wouldn't really benefit me.  If someone copies things one at a time as they need them then not having to authenticate each time would be nice, but to me it doesn't seem like the way it would usually happen.

Another thought I had was for the instructor in course B to be able to configure who could copy materials from that course.  The benefit of this approach is that people can share content with other instructors without having to give them permission access to the course.  A couple of methods come to mind:
- The instructor explicitly sets which courses can copy from that course, with options being
    - Any other course on this server
    - An explicit list of courses
    - Nobody
- The instructor sets an access key for their course, and an instructor in course A must enter that access key to copy from course B.

I have hacked some of this using symlinks.  If you symlink a subfolder of templates to another course then multiple courses can access those files.  The downside is that a change in one course can have effects on other courses.  We've also had multi-section courses where we symlinked the whole templates folder.  This eliminates the download/upload step since .def files will immediately be available in the other section, but gets messy when you have a number of WW courses sharing the same templates folder.  Of course this whole approach is kludgy and depends on def files, which shouldn't be necessary within a single WW server.

**Reply by Alex-Jordan** [reply]
All of this is a good exercise for thinking thoroughly about WW3 database structure. I've noticed that MyOpenMath has a structure in place for how you can explicitly release content to specific users or to the world at large. (Not sure if it allows you to release to groups like instructors within your institution.) It has tiers as well, like releasing a problem that you wrote to be used by others, but not releasing access to the underlying PG code. I feel that the OPL has failed to protect the rights of the contributors, and we can do better with WW3.
